presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PredicatePushDown.java

```diff
@@ -489,7 +489,10 @@ public PlanNode visitJoin(JoinNode node, RewriteContext<Expression> context)
                 output = new FilterNode(idAllocator.getNextId(), output, postJoinPredicate);
             }
 
-            output = new ProjectNode(idAllocator.getNextId(), output, Assignments.identity(node.getOutputSymbols()));
+            if (!node.getOutputSymbols().equals(output.getOutputSymbols())) {
+                output = new ProjectNode(idAllocator.getNextId(), output, Assignments.identity(node.getOutputSymbols()));
```

Comment:
If `node.getOutputSymbols().equals(output.getOutputSymbols())`, then there should be some prune identity projections rule that will remove it, right?
I don't understand value of this change. Please explain.

Reply:
We want to avoid creating new plan node if not needed in old `PlanOptimizer` execution. Otherwise you would need to always append iterative optimizer afterwards. Ideally it should be possible to spin predicate pushdown until something changes.

Reply:
> Ideally it should be possible to spin predicate pushdown until nothing changes so that all predicates propagate.

Is that your end goal? Are we making this step? 
You can always loop (PPD+pruning) instead of looping PPD alone.
Tis adds more logic into an already complex class. I want to be sure we're reaping benefits of that.

Reply:
> Is that your end goal? Are we making this step?

The goal was to make predicate pushdown stable and avoid regenerating new plan node if not needed which all optimizers do (including iterative) 

> Tis adds more logic into an already complex class. I want to be sure we're reaping benefits of that.

Some tests were fixed that changed predefined order of predicates so it's beneficial as planning is more stable (not necessarily with this commit)

Reply:
If it is all about symbol pruning here, maybe we should remove this projection at all and set output symbols in join node properly

Reply:
It's not always possible as cross join does not support pruning